### PR TITLE
Add width constraints for FlexibleSpaceBar.title in its expanded state, so that overflow of long titles can be handled

### LINKED ISSUE #14227: FlexibleSpaceBar title overflows on top of AppBar actions
This is a regression, it worked fine in the past. However, I don't remember the last good commit. It broke with one of the commits made in the past few weeks.

## Steps to Reproduce

Run the sample code, the bug should be visible immediately.

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(new Sample());
}

class Sample extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      home: new Scaffold(
        body: new CustomScrollView(
          slivers: <Widget>[
            new SliverAppBar(
              actions: <Widget>[
                new IconButton(
                  icon: new Icon(Icons.cancel),
                  onPressed: null,
                ),
                new IconButton(
                  icon: new Icon(Icons.cancel),
                  onPressed: null,
                ),
              ],
              flexibleSpace: const FlexibleSpaceBar(
                title: const Text(
                    "asdfkljsdfkljsdfkalskdfjalksdjflksjdflkasjdflkasjdfkljasdlfkj"),
                centerTitle: false,
              ),
            ),
          ],
        ),
      ),
    );
  }
}
```

## Logs

Running flutter run and flutter analyze did not output anything.

## Flutter Doctor

```
[✓] Flutter (on Mac OS X 10.13.2 17C88, locale en-NZ, channel master)
    • Flutter version 0.0.21-pre.294 at /Users/gnawf/flutter
    • Framework revision 8059aea365 (10 hours ago), 2018-01-22 18:12:39 -0800
    • Engine revision 171d032f86
    • Tools Dart version 2.0.0-dev.16.0
    • Engine Dart version 2.0.0-edge.93d8c9fe2a2c22dc95ec85866af108cfab71ad06

[✓] Android toolchain - develop for Android devices (Android SDK 26.0.3)
    • Android SDK at /usr/local/opt/android-sdk/
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-26, build-tools 26.0.3
    • ANDROID_HOME = /usr/local/opt/android-sdk/
    • Java binary at: /Library/Java/JavaVirtualMachines/jdk1.8.0_144.jdk/Contents/Home/bin/java
    • Java version Java(TM) SE Runtime Environment (build 1.8.0_144-b01)

[-] iOS toolchain - develop for iOS devices (Xcode 9.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 9.2, Build version 9C40b
    ✗ Your Mac needs to enabled for developer mode before using Xcode for the first time.
      Run 'sudo DevToolsSecurity -enable' to enable developer mode.
    ✗ libimobiledevice and ideviceinstaller are not installed. To install, run:
        brew install --HEAD libimobiledevice
        brew install ideviceinstaller
    ✗ ios-deploy not installed. To install:
        brew install ios-deploy
    • CocoaPods version 1.3.1

[✗] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.io/setup/#android-setup for detailed instructions).

[✓] IntelliJ IDEA Community Edition (version 2017.3.3)
    • Flutter plugin version 21.0.3
    • Dart plugin version 173.4301.22

[✓] Connected devices
    • iPhone 8 • 533B1ADF-56E5-4A63-9319-1F704E8845E0 • ios • iOS 11.2 (simulator)
```

### PR DESCRIPTION
**Pull request by perclasson**
Add width constraints for FlexibleSpaceBar.title in its expanded state, so that overflow of long titles can be handled.

See [this dartpad for reproducible example of the issue](https://dartpad.dartlang.org/68be44f60cb1b852a8c24cbbb539f359).

## Description

Add width constraints for FlexibleSpaceBar.title in its expanded state, so that overflow of long titles can be handled.

Before:

After:

Example with larger text scale factor:

## Related Issues

Closes https://github.com/flutter/flutter/issues/14227

## Tests

I added the following tests:

- FlexibleSpaceBar sets width constraints for the title

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I signed the [CLA].
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] All existing and new tests are passing.
- [X] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [X] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [X] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/test/material/flexible_space_bar_test.dart [hunk 1]
```diff
@@ -117,6 +119,67 @@ void main() {
...
+    final FlexibleSpaceBarSettings customSettings = FlexibleSpaceBar.createSettings(
+      currentExtent: maxExtent,
+      minExtent: minExtent,
+      maxExtent: maxExtent,
+      child: AppBar(
+        flexibleSpace: FlexibleSpaceBar(
+          titlePadding: EdgeInsets.zero,
+          title: Container(
+            height: height,
+            child:  Text(
```
Comment (pennzht) [inline packages/flutter/test/material/flexible_space_bar_test.dart]:
```suggestion
            child: Text(
```

**Review by pennzht — APPROVED** [review]

**Comment by pennzht** [comment]
You can add a code snippet to the PR description to clarify what issue is being solved.

File: packages/flutter/test/material/flexible_space_bar_test.dart [hunk 2]
```diff
@@ -117,6 +119,67 @@ void main() {
...
+      currentExtent: maxExtent,
+      minExtent: minExtent,
+      maxExtent: maxExtent,
+      child: AppBar(
+        flexibleSpace: FlexibleSpaceBar(
+          titlePadding: EdgeInsets.zero,
+          title: Container(
+            height: height,
+            child: Text(
+              'X' * 2000,
```
Comment (rami-a) [inline packages/flutter/test/material/flexible_space_bar_test.dart]: Had no idea you could do this, neat!

**Review by rami-a — APPROVED** [review]
LGTM

File: packages/flutter/test/material/flexible_space_bar_test.dart [hunk 3]
```diff
@@ -117,6 +119,67 @@ void main() {
...
   });
 
+  // This is a regression test for https://github.com/flutter/flutter/issues/14227
+  testWidgets('FlexibleSpaceBar sets width constraints for the title', (WidgetTester tester) async {
+    const double width = 300;
+    const double height = 100;
+    const double minExtent = 100;
+    const double maxExtent = 200;
+
+    final FlexibleSpaceBarSettings customSettings = FlexibleSpaceBar.createSettings(
```
Comment (HansMuller) [inline packages/flutter/test/material/flexible_space_bar_test.dart]: This test doesn't need to be so complicated, it should really be a minimal version of the test case provided with https://github.com/flutter/flutter/issues/14227.

File: packages/flutter/test/material/flexible_space_bar_test.dart [hunk 4]
```diff
@@ -117,6 +119,67 @@ void main() {
...
+                  ),
+                ),
+              ],
+            ),
+          ),
+        ),
+      ),
+    );
+
+    // The title is scaled and transformed to be 1.5 times bigger, when the
```
Comment (HansMuller) [inline packages/flutter/test/material/flexible_space_bar_test.dart]: Why not just verify that `tester.getRect(find.byType(Text))` is what you'd expect?

Comment (perclasson) [reply packages/flutter/test/material/flexible_space_bar_test.dart]: Good idea, that made it simpler.

File: packages/flutter/test/material/flexible_space_bar_test.dart (hunk 3 above)
Comment (perclasson) [reply packages/flutter/test/material/flexible_space_bar_test.dart]: I simplified the test now.

File: packages/flutter/test/material/flexible_space_bar_test.dart [hunk 5]
```diff
@@ -117,6 +117,49 @@ void main() {
     expect(clipRect.size.height, minExtent);
   });
 
+  // This is a regression test for https://github.com/flutter/flutter/issues/14227
+  testWidgets('FlexibleSpaceBar sets width constraints for the title', (WidgetTester tester) async {
```
Comment (HansMuller) [inline packages/flutter/test/material/flexible_space_bar_test.dart]: This version of the test looks much better.

File: packages/flutter/test/material/flexible_space_bar_test.dart [hunk 6]
```diff
@@ -117,6 +117,49 @@ void main() {
...
+              );
+            }
+          ),
+        ),
+      ),
+    );
+
+    // The title is scaled and transformed to be 1.5 times bigger, when the
+    // FlexibleSpaceBar is fully expanded, thus we expect the width to be
+    // 1.5 times smaller than the full width.
```
Comment (HansMuller) [inline packages/flutter/test/material/flexible_space_bar_test.dart]: Please add a sentence or two that explains where the -30 and -10 adjustments (top,bottom) come from.

**Review by HansMuller — APPROVED** [review]
LGTM

File: packages/flutter/test/material/flexible_space_bar_test.dart (hunk 6 above)
Comment (perclasson) [reply packages/flutter/test/material/flexible_space_bar_test.dart]: Done!

**Pull request merged**